File: JEDI/bin/runAssimilation.py
#!/bin/env python3

import sys
import os
import shutil
import yaml
from datetime import datetime, timedelta
import subprocess

import leadtime

# Get experiment yaml file path from comand line
exp_config_file = sys.argv[1]

# Get the analysis time from the command line
analysis_time_str = sys.argv[2]
analysis_time = datetime.strptime(analysis_time_str, "%Y-%m-%dT%H:%M:%SZ")

# Extract path to experiment configuration
exp_config_path = os.path.dirname(os.path.abspath(exp_config_file))

# Load the experiment configuration
with open(exp_config_file, 'r') as file:
    exp_config = yaml.safe_load(file)

# Get the assimilation type (3d, 4d)
assimilation_type = exp_config['assimilation']['type']
assimilation_algorithm = exp_config['assimilation']['algorithm']
    
# Load the assimilation config template
with open(f'{exp_config_path}/{assimilation_type}.{assimilation_algorithm}.yaml') as file:
    assim_config = yaml.safe_load(file)

# Get the experiment path
exp_path = f"{exp_config['experiment']['path']}/{exp_config['experiment']['name']}"

# Create the assimilation output directory
fcst_path = f'{exp_path}/forecasts/{analysis_time_str}/on'
if not os.path.exists(fcst_path):
    os.makedirs(fcst_path)

# Set up the assimilation window configuration
window_begin = analysis_time + timedelta(0, leadtime.fcst_to_seconds(exp_config['assimilation']['window begin']))
assim_config['cost function']['window begin'] = window_begin.strftime("%Y-%m-%dT%H:%M:%SZ")
assim_config['cost function']['window length'] = exp_config['assimilation']['window length']

# Get previous cycle time
prev_cycle = analysis_time - timedelta(0, leadtime.fcst_to_seconds(exp_config['experiment']['cycle frequency']))
prev_cycle_str = prev_cycle.strftime("%Y-%m-%dT%H:%M:%SZ")

# Set up the background file configuration
if assimilation_type == "4dvar":
    # 4DVar background files must be at the beginning of the assimilation window
    bkg_offset = leadtime.fcst_to_seconds(exp_config['assimilation']['window begin']) + \
                 leadtime.fcst_to_seconds(exp_config['experiment']['cycle frequency'])
else:
    # 3DVar background files must be at the center of the assimilation window
    bkg_offset = leadtime.fcst_to_seconds(exp_config['assimilation']['window begin']) + \
                 leadtime.fcst_to_seconds(exp_config['experiment']['cycle frequency']) + \
                 (leadtime.fcst_to_seconds(exp_config['assimilation']['window length']) // 2)
bkg_date = analysis_time - \
           timedelta(0, leadtime.fcst_to_seconds(exp_config['experiment']['cycle frequency'])) + \
           timedelta(0, bkg_offset)
bkg_date_str = bkg_date.strftime("%Y-%m-%dT%H:%M:%SZ")
assim_config['cost function']['background']['date'] = bkg_date_str
assim_config['cost function']['background']['filename'] = f'{exp_path}/forecasts/{prev_cycle_str}/on/forecast.fc.{prev_cycle_str}.{leadtime.seconds_to_fcst(bkg_offset)}.nc'

# Set up the obs file configuration
observers = assim_config['cost function']['observations']['observers']
for observer in observers:
    obs_space = observer['obs space']
    obs_space['obsdatain']['engine']['obsfile'] = f'{exp_path}/obs/qg.truth.{assimilation_type}.{analysis_time_str}.nc'
    obs_space['obsdataout']['engine']['obsfile'] = f'{fcst_path}/qg.{assimilation_type}.{analysis_time_str}.nc'
    
# Set up assimilation output configuration
assim_config['output']['datadir'] = fcst_path
iterations_config = assim_config['variational']['iterations']
for iteration in iterations_config:
    print(iteration)
    iteration['online diagnostics']['increment']['datadir'] = fcst_path
assim_config['final']['increment']['output']['datadir'] = fcst_path

# Write out the forecast configuration
if (not os.path.exists(f'{exp_path}/yaml')):
    os.makedirs(f'{exp_path}/yaml')
with open(f"{exp_path}/yaml/{assimilation_type}.{assimilation_algorithm}.{analysis_time_str}.yaml", 'w') as file:
    yaml.dump(assim_config, file)
        
# Run the forecast
with open(f'{fcst_path}/{assimilation_type}.{assimilation_algorithm}.{analysis_time_str}.log', 'w') as logfile:
    subprocess.run([f"{exp_config['jedi path']}/bin/qg_4dvar.x", f'{exp_path}/yaml/{assimilation_type}.{assimilation_algorithm}.{analysis_time_str}.yaml'], stdout = logfile, stderr = sys.stdout)

File: JEDI/bin/runAssimilation.rb
#!/usr/bin/ruby

# Get the base directory
baseDir=File.expand_path('../..',__FILE__)

# Add bin directory to require search path
$LOAD_PATH.unshift "#{baseDir}/bin"

# Load libs
require 'yaml'
require 'date'
require 'fileutils'
require 'leadTime.rb'

# Load the experiment config
expConfig = YAML.load_file(ARGV[0])

# Get the assimilation type (3d, 4d)
assimilationType = expConfig['Assimilation']['Type']
assimilationAlgorithm = expConfig['Assimilation']['Algorithm']

# Get the analysis time
anlString=ARGV[1]
anlTime=DateTime.parse(anlString).to_time

# Load the base assimilation config template
assimConfig = YAML.load_file("#{baseDir}/yaml/#{assimilationType}.#{assimilationAlgorithm}.yaml")

# Create experiment assimilation output directory
expDir = "#{baseDir}/experiments/#{expConfig['Experiment']['Name']}"
fcstDir = "#{expDir}/forecasts/#{anlString}/assimilation_on"
FileUtils.mkdir_p("#{fcstDir}")

# Set up the assimilation window configuration
assimConfig["cost_function"]["window_begin"] = (anlTime + LeadTime.fcstToSeconds(expConfig['Assimilation']['Window Begin'])).strftime("%Y-%m-%dT%H:%M:%SZ")
assimConfig["cost_function"]["window_length"] = expConfig['Assimilation']['Window Length']

# Get the previous cycle
prevCycle = (anlTime - LeadTime.fcstToSeconds(expConfig['Experiment']['Cycle Frequency'])).strftime("%Y-%m-%dT%H:%M:%SZ")

# Set up the background file configuration
# 4DVar background files must be at the beginning of the assimilation window
if (expConfig['Assimilation']['Type'] == '4dvar')
  bkgOffset = LeadTime.fcstToSeconds(expConfig['Assimilation']['Window Begin']) + LeadTime.fcstToSeconds(expConfig['Experiment']['Cycle Frequency'])
# 3DVar background files must be in the center of the assimilation window
else
  bkgOffset = LeadTime.fcstToSeconds(expConfig['Assimilation']['Window Begin']) + LeadTime.fcstToSeconds(expConfig['Experiment']['Cycle Frequency']) + (LeadTime.fcstToSeconds(expConfig['Assimilation']['Window Length']) / 2)
end
bkgDate = (anlTime - LeadTime.fcstToSeconds(expConfig['Experiment']['Cycle Frequency']) + bkgOffset).strftime("%Y-%m-%dT%H:%M:%SZ")
assimConfig["cost_function"]["Jb"]["Background"]["state"].each_with_index do |bkg, i|
  assimConfig["cost_function"]["Jb"]["Background"]["state"][i]["filename"] = "#{expDir}/forecasts/#{prevCycle}/assimilation_on/test.fc.#{prevCycle}.#{LeadTime.secondsToFcst(bkgOffset)}"
  assimConfig["cost_function"]["Jb"]["Background"]["state"][i]["date"] = bkgDate
end

# Set up the background covariance matrix configuration
assimConfig["cost_function"]["Jb"]["Covariance"]["date"] = anlString
assimConfig["cost_function"]["Jb"]["Covariance"]["covariance"] = "L95Error"

# Set up the observation file configuration
assimConfig["cost_function"]["Jo"]["ObsTypes"].each_with_index do |obsType, i|
  assimConfig["cost_function"]["Jo"]["ObsTypes"][i]["ObsData"]["ObsDataIn"]["filename"] = "#{expDir}/obs/l95.truth.#{assimilationType}.#{anlString}.obt"
  assimConfig["cost_function"]["Jo"]["ObsTypes"][i]["ObsData"]["ObsDataOut"]["filename"] = "#{fcstDir}/l95.#{assimilationType}.#{anlString}.obt"
end

# Set up output directory configuration
assimConfig["output"]["datadir"] = fcstDir

# Write out updated assimilation config
FileUtils.mkdir_p("#{expDir}/yaml")
File.open("#{expDir}/yaml/#{assimilationType}.#{assimilationAlgorithm}.#{anlTime.strftime("%Y-%m-%dT%H:%M:%SZ")}.yaml", "w") { |f| YAML.dump(assimConfig, f) }

# Save the experiment config
File.open("#{expDir}/yaml/#{expConfig['Experiment']['Name']}.yaml", "w") { |f| YAML.dump(expConfig, f) }

# Run the assimilation
system("#{expConfig['JEDI Path']}/bin/l95_4dvar.x #{expDir}/yaml/#{assimilationType}.#{assimilationAlgorithm}.#{anlTime.strftime("%Y-%m-%dT%H:%M:%SZ")}.yaml > #{fcstDir}/runAssimilation.stdout")

if $?.exitstatus != 0
  raise "runAssimilation failed!  See #{fcstDir}/runAssimilation.stdout"
end
